pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java

```diff
@@ -255,19 +255,15 @@ private boolean verifyChecksum(ByteBuf headersAndPayload) {
     }
 
     private void startPublishOperation(int batchSize, long msgSize) {
-        // A single thread is incrementing/decrementing this counter, so we can use lazySet which doesn't involve a mem
-        // barrier
-        pendingPublishAcksUpdater.lazySet(this, pendingPublishAcks + 1);
+        // Since different threads can start and complete the publish operation, we need a strong concurrency guarantee
+        pendingPublishAcksUpdater.incrementAndGet(this);
         // increment publish-count
         this.getTopic().incrementPublishCount(batchSize, msgSize);
     }
 
     private void publishOperationCompleted() {
-        long newPendingPublishAcks = this.pendingPublishAcks - 1;
-        pendingPublishAcksUpdater.lazySet(this, newPendingPublishAcks);
-
         // Check the close future to avoid grabbing the mutex every time the pending acks goes down to 0
-        if (newPendingPublishAcks == 0 && !closeFuture.isDone()) {
+        if (pendingPublishAcksUpdater.decrementAndGet(this) <= 0 && !closeFuture.isDone()) {
```

Comment:
For the sake or readability I would push this decrement operation out of the 'if, one line above.
Personally I find difficult to read 'if' conditions with multiple cases and some expression is mutating the variables.
Separating state mutations from tests I easier to understand